File: web/src/entities/reactorCore.ts
import { Direction } from '../constants.js';
import { ImmutableVec2 } from '../math.js';
import { Entity } from '../entity.js';
import { PlayerSkill } from '../playerSkill.js';
import { Resource } from '../resource.js';
import type { Animation } from '../media/animation.js';

export class ReactorCore extends Entity {
  private readonly animation: Animation;
  private frame = 0;

  constructor() {
    super();
    this.setSize(new ImmutableVec2(30, 30));
    this.animation = Resource.getAnimation('data/images/core.bmp', 3);
  }

  getLayer(): number { return 1; }

  override draw(ctx: CanvasRenderingContext2D, offsetX: number, offsetY: number, _layer: number): void {
    const pos = new ImmutableVec2(this.getPosition())
      .subtract(new ImmutableVec2(this.animation.getFrameWidth(), this.animation.getFrameHeight()).divide(2))
      .add(new ImmutableVec2(offsetX, offsetY));
    this.animation.drawFrame(ctx, Math.floor(this.frame / 5), Math.round(pos.x), Math.round(pos.y));
  }

  override update(): void {
    this.myVelocity.set(0, this.myVelocity.y + 6.0);
    const bumps = this.moveWithCollision();
    if (bumps.has(Direction.UP) || bumps.has(Direction.DOWN)) {
      this.myVelocity.set(this.myVelocity.x, this.myVelocity.y * 0.8);
    }

    if (this.myRoom.getHero().Collides(this)) {
      if (this.myRoom.getHero().gotCore()) {
        PlayerSkill.playerDidSomethingClever(1.0, 0.75);
        this.myRoom.setCompleted();
        this.remove();
      }
    }
    this.frame++;
  }
}

File: wii/src/entities/reactor_core.cpp
#include "entities/reactor_core.h"
#include "entities/hero.h"
#include "resource.h"
#include "media/animation.h"
#include "room.h"
#include "player_skill.h"
#include <cmath>

ReactorCore::ReactorCore() {
    setSize(Vec2(30, 30));
    animation_ = Resource::getAnimation("data/images/core.bmp", 3);
}

void ReactorCore::draw(SDL_Renderer* renderer, int offsetX, int offsetY, int /*layer*/) {
    int x = static_cast<int>(std::round(position_.x - animation_->getFrameWidth() / 2.0f + offsetX));
    int y = static_cast<int>(std::round(position_.y - animation_->getFrameHeight() / 2.0f + offsetY));
    animation_->drawFrame(renderer, frame_ / 5, x, y);
}

void ReactorCore::update() {
    velocity_.y += 6.0f;
    auto bumps = moveWithCollision();
    if (bumps.count(Direction::UP) || bumps.count(Direction::DOWN)) {
        velocity_.y *= 0.8f;
    }

    if (room_->getHero()->Collides(*this)) {
        if (room_->getHero()->gotCore()) {
            PlayerSkill::playerDidSomethingClever(1.0f, 0.75f);
            room_->setCompleted();
            remove();
        }
    }
    frame_++;
}
